### ISSUE TITLE
Unhandled CicChip(-1) in First DMA 

### ISSUE THREAD

**Issue by theboy181** [project64/project64#61]
Unhandled CicChip(-1) in First DMA error in pj64 when trying to load Aleck64 ROMS

Some ROMS will load but into test menu (Eleven Beat). Missing RDB entry for proper testing too.

**Comment by death-droid** [comment]
This is due to Aleck64 roms having a different CiC that Project64 does not recognize. However from what I've read 6102 may allow them to work.

Might make sense for us to actually use a default Cic rather than causing it to set to -1 and cause the emulator to be useless even if the ROM may still run under another CIC

**Comment by death-droid** [comment]
Aleck64 ROMS use a CiC chip that is not support by Project 64.  CIC-NUS-6102 may work. Otherwise we need to implement CIC-NUS-5101. I can assume once that's resolved there is probably a bunch other problems will be ran into.

This is probably more up Zilmar's alley. 
@project64

**Comment by theboy181** [comment]
These ROMS are loading fine in Mupen64 so I am unsure if it supports the CIC's then.

**Comment by project64** [comment]
http://wiki.arcadeotaku.com/w/Seta_Aleck_64

I guess it just needs some tweaks to get them to run? I have never tried any so I have no idea

**Comment by theboy181** [comment]
I managed to get all but one of the games running in PJ64. 

Basically it's just missing from the RDB. Need to set memory to 8 instead if 4 and I am unsure if there are other settings that should be tweaked. 

You still get the unidentified CIC error. Not sure if you just want to skip that message or support the CIC. 

Is there a guide on what all the settings for a ROM do and how they should be set?

It would be really benificial if these games could be identified and governed outside of the RDB. This would make loading PD ROMS a little less time consuming.

**Comment by theboy181** [comment]
If you want I can get you the patches for the ROMS so you can further test. PM me on the forums.

**Comment by cxd4** [comment]
What's the only ROM left that RDB configuration doesn't get it to work?
Aleck64, was it?

Is that a public domain image I can find somewhere or something?  Does it work on other emu?

If it's difficult to troubleshoot openly I can just PM or talk about it over irc, but maybe irc would be quicker.

**Comment by theboy181** [comment]
Tower&Shaft I cant seem to get it to load. Unsure what the correct settings are supposed to be.

Its a ALECK64 modified ROM and its running in Mupen64 iirc

Is there a function that adds the ROM settings to the RDB so that I can get the CRC's added to it for editing?

**Comment by cxd4** [comment]
> Is there a function that adds the ROM settings to the RDB so that I can get the CRC's added to it for editing?

Don't think so, apart from just copy-pasting RDB entries and renaming the relevant bits of it like said CRCs.

> When are you usually on IRC? I don't see you there much.

If I'm currently connected to the Internet, then I am probably on IRC.  Otherwise, I have a BNC to help me see offline messages.

**Comment by cxd4** [comment]
Okay looking into this more, it falls under PI DMA (wow this again?).

The func is referenced just once, when in PI_DMA_WRITE under `dma.cpp`:
```c
		if (!g_System->DmaUsed())
		{
			g_System->SetDmaUsed(true);
			OnFirstDMA(); 
		}
```
error happens from said function at:
```c
void CDMA::OnFirstDMA (void) {
	switch (g_Rom->CicChipID()) {
	case 1: *(DWORD *)&((g_MMU->Rdram())[0x318]) = g_MMU->RdramSize(); break;
	case 2: *(DWORD *)&((g_MMU->Rdram())[0x318]) = g_MMU->RdramSize(); break;
	case 3: *(DWORD *)&((g_MMU->Rdram())[0x318]) = g_MMU->RdramSize(); break;
	case 5: *(DWORD *)&((g_MMU->Rdram())[0x3F0]) = g_MMU->RdramSize(); break;
	case 6: *(DWORD *)&((g_MMU->Rdram())[0x318]) = g_MMU->RdramSize(); break;
	default: g_Notify->DisplayError("Unhandled CicChip(%d) in first DMA",g_Rom->CicChipID());
	}
}
```
I'll look more into it when I wake up.  The UI interface to pj64 doesn't recognize the CIC chip type either.  I can say for now that both 1964 and Mupen64 recompilers load the ROM easily and without issue, so it should be supported internally within a future version of Project64.

**Comment by death-droid** [comment]
Mupen64 i know gets around issues like this as instead of setting the CicChip to like -1.
It sets it to CIC_NUS_6102. Probably better for the end user if we do the same, even if we only do it when the debugger isn't active. (or not building under debug)

**Comment by cxd4** [comment]
Apparently so.

In `mupen64/r4300/dma.c`:
```c
   if ((debug_count+Count) < 0x100000)
     {
	switch(CIC_Chip)
	  {
	   case 1:
	   case 2:
	   case 3:
	   case 6:
	     rdram[0x318/4] = 0x800000;
	     break;
	   case 5:
	     rdram[0x3F0/4] = 0x800000;
	     break;
	  }
     }
```
Basically it's identical to how Project64 does it in func `OnFirstDMA`, except that there is no `default` case, due to what you said--it defaults to 6102 CIC.

And 1964 seems to inherently apply an old method PJ64 used to deduce the CIC straight from the header...wasn't able to find anything about using a checksum in it.

**Comment by cxd4** [comment]
Simply appending this at the end of `CN64Rom::CalculateCicChip` is enough to take care of the first error.
```c
    if (m_CicChip == CIC_UNKNOWN)
        m_CicChip = CIC_NUS_6102;
```
However the real issue behind why this game doesn't work only in Project64, seems to lie elsewhere.

The real issue actually has little to do with the CIC boot-code.  That was just a warning popped up from pj.
The real issue is a separate breakpoint from `Memory Virtual Mem.cpp`.

The basic error flow from within interpreter seems to be:

1.  `CN64System::StartEmulation`, `MemoryFilter` begin, `MemoryFilter` end
2.  `CInterpreterCPU::ExecuteCPU`, `MemoryFilter` begin, {... basic interpreter loop begins}
3.  `CN64System::RefreshScreen`, `MemoryFilter` begin, `MemoryFilter` end
4.  repeat the above thousands of times, for thousands of vertical interrupts
5.  New `VI_CONTROL_REG` (message from my plugin):  0x0000311E
6.  New `VI_CONTROL_REG` (message from my plugin):  0x00013012
7.  more heaps of vertical interrupts (had to set VI skip to 255 in my plugin to not spam so much)
8.  `CN64System::RunRSP`, `MemoryFilter` begin, `MemoryFilter` end
9.  another intermediate vertical interrupt
10.  `CN64System::RunRSP`, `MemoryFilter` begin, `MemoryFilter` end
11.  `CN64System::RunRSP`, `MemoryFilter` begin, `MemoryFilter` end
12.  BREAKPOINT:  `N64 System\Mips\Memory Virtual Mem.cpp`, line 1925
13.  `MemoryFilter` end (of basic interpreter loop, see item number 2.)
14.  emu dies

`Interpreter CPU.cpp`, line 404:
```c
	} __except( g_MMU->MemoryFilter( GetExceptionCode(), GetExceptionInformation()) ) {
```
which calls up until `Memory Virtual Mem.cpp`, see next post ...

**Comment by cxd4** [comment]
... which is a breakpoint inside of `LH_NonMemory`, which is being called exactly when and why?

On line 1608 of `int CMipsMemoryVM::MemoryFilter(DWORD dwExptCode ...)`:
```c
	BYTE * TypePos = (unsigned char *)lpEP->ContextRecord->Eip;
```
and starting at line 1720 of the same func, after that declaration:
```c
	switch(*TypePos) {
/* ... irrelevant cases here */
	case 0x66:
		switch(*(TypePos + 1)) {
		case 0x8B:g_Notify->DisplayError("y");
			if (!LH_NonMemory(MemAddress,Reg,FALSE)) {
				if (g_Settings->LoadDword(Debugger_ShowUnhandledMemory)) {
					g_Notify->DisplayError("Failed to half word\n\nMIPS Address: %X\nX86 Address",
						(char *)exRec.ExceptionInformation[1] - (char *)m_RDRAM,
						*(unsigned char *)lpEP->ContextRecord->Eip);
				}
			}g_Notify->DisplayError("z");
			lpEP->ContextRecord->Eip = (DWORD)ReadPos;
			return EXCEPTION_CONTINUE_EXECUTION;	
```
So Project64 pops up the "y" DisplayError string I told it to show, then does the breakpoint at step 13 in my above post, then prints "z" on the screen with another MessageBox.

This is the reason that `LH_NonMemory` fails within that switch case:
```c
int CMipsMemoryVM::LH_NonMemory ( DWORD PAddr, DWORD * Value, int/* SignExtend*/ ) 
{
	if (PAddr < 0x800000)
	{
		* Value = 0;
		return true;
	}

	if (PAddr >= 0x10000000 && PAddr < 0x16000000) 
	{
		g_Notify->BreakPoint(__FILE__,__LINE__);
	}
```
In other words, an invalid physical address (not mapped to the respective portion of RCP memory) was being passed, which caused this ROM to fail.  Next question is why this didn't happen in a working CPU flow in the other emulators...

**Comment by cxd4** [comment]
Replacing the breakpoint as follows:
```c
    if (PAddr >= 0x10000000 && PAddr < 0x16000000) 
    {
        //g_Notify->BreakPoint(__FILE__,__LINE__);
        g_Notify -> DisplayError("o shit lel\naddr:  %lu\n", PAddr);
    }
```
Shows that the address constantly cycles back and forth between these 2 values, forever as far as I can tell:  0x10000C06 and 0x10000C0A.

The audio is very loud--enough to rape my ears if I don't turn down the audio plugin volume.  On Jabo's DirectSound LLE there is no audio; Azimer's plugin as LLE there is audio...I find that interesting.

This happens whether I put `return true;` or `return false;` at the end of that block.
So I think that this is an unhandled case overlooked somewhere in the interpreter core.

**Comment by death-droid** [comment]
Can this be closed?
If its a specific game that doesnt work with recent changes than a new issue should be made, or if the game fails to initialize properly with the CIC_UNKNOWN

@theboy181 @project64 @cxd4

**Comment by cxd4** [comment]
Fair point.

The actual issue is a TLB regression since after Project64 1.6, so I would close this issue and recreate it @theboy181 .

**Comment by theboy181** [comment]
Can someone who understands the actual issue open it up please?

**Comment by theboy181** [comment]
Still getting CIC error on ROM LOAD

**Comment by cxd4** [comment]
It's not an error.  It's a debugger message.

It's handled the same way as it is in Mupen64; you see an "error" because you enabled debugging.

**Comment by theboy181** [comment]
Thanks. I thought it was strange that it was working then stopped. I guess I should turn off debug.

**Comment by theboy181** [comment]
Thanks. I thought it was strange that it was working then stopped. I guess I should turn off debug.
